cargo-dist/src/host.rs

```diff
@@ -0,0 +1,240 @@
+                // Only update them if they don't already have hosting
+                // if create_hosting is set, then consider all entries out of date
+                // and needing refreshing (this is only set by `cargo dist host create`)
+                let needs_hosting = create_hosting
+                    || self
+                        .manifest
+                        .release_by_name(&name)
+                        .map(|r| r.hosting.is_none())
+                        .unwrap_or(true);
+                if needs_hosting {
+                    Some((name, version))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        // If everything was already provided by a merged dist-manifest, don't redo it
+        if releases_without_hosting.is_empty() {
+            return Ok(());
+        }
+
+        match hosting.hosts {
+            HostingStyle::Axodotdev => {
+                // Ask The Abyss For Hosting, or mock the result
+                let packages = releases_without_hosting
+                    .iter()
+                    .map(|(name, _version)| name.clone());
+
+                let artifact_sets = if create_hosting {
+                    let abyss =
+                        gazenot::Gazenot::into_the_abyss(&hosting.source_host, &hosting.owner)?;
+                    tokio::runtime::Handle::current()
+                        .block_on(abyss.create_artifact_sets(packages))?
+                } else {
+                    packages.map(gazenot::ArtifactSet::mock).collect()
+                };
+
+                // Store the results so other machines can use it
+                for ((name, version), set) in
+                    releases_without_hosting.into_iter().zip(artifact_sets)
+                {
+                    assert_eq!(
+                        name, set.package,
+                        "gazenot got confused about package names..."
+                    );
+                    self.manifest.ensure_release(name, version).hosting = Some(Hosting::Abyss(set));
+                }
+            }
+            HostingStyle::Github => {
+                // CI currently impls this for us, all we need to know is the URL to download from
+                let repo_url = &hosting.repo_url;
+                for (name, version) in releases_without_hosting {
+                    let tag = &announcing.tag;
+                    self.manifest.ensure_release(name, version).hosting =
+                        Some(Hosting::Github(cargo_dist_schema::GithubHosting {
+                            artifact_download_url: format!("{repo_url}/releases/download/{tag}"),
+                        }))
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn check_hosting(_dist: &DistGraph, _manifest: &DistManifest, _abyss: &Gazenot) -> DistResult<()> {
+    // FIXME: implement a ping/whoami API to check the Abyss client is working
```

Comment:
Is this getting done pre- or post-prerelease?

Reply:
post, liv is out.

Reply:
👍